core/jvmMain/src/kotlinx/serialization/internal/Caching.kt

```diff
@@ -101,10 +103,37 @@ private class ConcurrentHashMapParametrizedCache<T>(private val compute: (KClass
 
 private class CacheEntry<T>(@JvmField val serializer: KSerializer<T>?)
 
+private class KTypeWrapper(val origin: KType): KType {
+    override val annotations: List<Annotation> = origin.annotations
+    override val arguments: List<KTypeProjection> = origin.arguments
+    override val classifier: KClassifier? = origin.classifier
+    override val isMarkedNullable: Boolean = origin.isMarkedNullable
+
+    override fun equals(other: Any?): Boolean {
+        if (other == null) return false
+        if (origin != other) return false
+
+        if (classifier is KClass<*>) {
+            val otherClassifier = (other as? KType)?.classifier
+            if (otherClassifier == null || otherClassifier !is KClass<*>) {
+                return false
+            }
+            return classifier.java == otherClassifier.java
```

Comment:
`KClassImpl` uses `javaObjectType` instead of `java`: https://github.com/JetBrains/kotlin/blob/bfa3f89aeb727518703f0a167153cb048724a6d1/core/reflection.jvm/src/kotlin/reflect/jvm/internal/KClassImpl.kt#L293

I'm not sure why, but maybe this difference is important

Reply:
As far as I understand, `javaObjectType` also handles primitives, but in our case primitives cannot parameterize types.